src/IO/ParallelReadBuffer.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include <IO/BufferWithOwnMemory.h>
+#include <IO/ReadBuffer.h>
+#include <IO/SeekableReadBuffer.h>
+#include <Common/ArenaWithFreeLists.h>
+#include <Common/ThreadPool.h>
+
+namespace DB
+{
+
+/**
+ * Reads from multiple ReadBuffers in parallel.
+ * Preserves order of readers obtained from ReadBufferFactory.
+ *
+ * It consumes multiple readers and yields data from them in order as it passed.
+ * Each working reader save segments of data to internal queue.
+ *
+ * ParallelReadBuffer in nextImpl method take first available segment from first reader in deque and fed it to user.
+ * When first reader finish reading, they will be removed from worker deque and data from next reader consumed.
+ *
+ * Number of working readers limited by max_working_readers.
+ */
+class ParallelReadBuffer : public SeekableReadBufferWithSize
+{
+private:
+    /// Blocks until data occurred in the first reader or this reader indicate finishing
+    /// Finished readers removed from queue and data from next readers processed
+    bool nextImpl() override;
+
+    class Segment : private boost::noncopyable
+    {
+    public:
+        Segment(size_t size_, SynchronizedArenaWithFreeLists * arena_) : arena(arena_), m_data(arena->alloc(size_)), m_size(size_) { }
+
+        Segment() = default;
+
+        Segment(Segment && other) noexcept : arena(other.arena)
+        {
+            std::swap(m_data, other.m_data);
+            std::swap(m_size, other.m_size);
+        }
+
+        Segment & operator=(Segment && other) noexcept
+        {
+            arena = other.arena;
+            std::swap(m_data, other.m_data);
+            std::swap(m_size, other.m_size);
+            return *this;
+        }
+
+        ~Segment()
+        {
+            if (m_data)
+            {
+                arena->free(m_data, m_size);
+            }
+        }
+
+        auto data() const noexcept { return m_data; }
+        auto size() const noexcept { return m_size; }
+
+    private:
+        SynchronizedArenaWithFreeLists * arena{nullptr};
+        char * m_data{nullptr};
+        size_t m_size{0};
+    };
+
+public:
+    struct Range
+    {
+        size_t from;
+        size_t to;
+    };
+
+    using ReaderWithRange = std::pair<ReadBufferPtr, Range>;
```

Comment:
May be a struct? Also this read buffer already has `getRemainingReadRange` method (all seekable read buffers have). May be you can reuse it instead.

Reply:
Ok, though all read buffers which we are interested in are inherited from `SeekableReadBuffer`, and also `getRemainingReadRange` can be moved from `SeekableReadBuffer` to `ReadBuffer` if needed. But current way is also ok.

Reply:
You are right, we use only SeekableReadBuffer for now, which is also true for `S3` later on.
It's better for me to use structs from it (`Range`) and the method.